### PR TITLE
feat: assertAndProcessMemo leverage noUncheckedIndexedAccess strictness

### PR DESCRIPTION
**Pull request by gomesalexandre**
## Description

We've been burned in `assertAndProcessMemo` by `String.prototype.split()` returning an array of strings, while it's really returning `(string|undefined)[]`, see e.g https://github.com/shapeshift/web/pull/6809.

While adding `noUncheckedIndexedAccess` isn't a reasonable option (i.e `Found 556 errors in 119 files.`), we can at least add it temporary and add the required strictness to the offenders here, which this PR does.

## Pull Request Type

- [ ] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [ ] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [x] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

N/A

Low, though being more strict comes with its own risks - confirm the checks are sane

> What protocols, transaction types or contract interactions might be affected by this PR?

## Testing

- Do a sanity pass of THOR Txs and confirm they are happy (no need to broadcast - set a breakpoint in`useSendThorTx`'s `executeTransaction` at the very first line of its body and ensure memos still look sane

### Engineering

- Ensure checks are sane
- Revert the revert of the tsconfig option efa3b15657cafa6debfa27202962b2e64e91189c
- Ensure there are no TS errors in the `memo.ts` file

### Operations

- ☝🏽

### CODE REVIEW

File: src/lib/utils/thorchain/memo.ts [hunk 1]
```diff
@@ -2,43 +2,60 @@ import { bn } from 'lib/bignumber/bignumber'
...
 
 // Disabling until we validate further, as :MINOUT is optional in the quote response
 // const assertMemoHasMinOut = (minOut: string, memo: string) => {
 //   if (!minOut) throw new Error(`minOut is required in memo: ${memo}`)
 // }
 
-const assertIsValidLimit = (limit: string, memo: string) => {
-  assertMemoHasLimit(limit, memo)
+const assertIsValidLimit = (limit: string | undefined, memo: string) => {
+  if (!assertMemoHasLimit(limit, memo)) return
```
Comment (gomesalexandre) [inline src/lib/utils/thorchain/memo.ts:52]: Same here this isn't really returning - this is an assertion, so will throw in the sad path, *not* return false

File: src/lib/utils/thorchain/memo.ts [hunk 2]
```diff
@@ -71,6 +95,8 @@ const assertIsValidBasisPoints = (basisPoints: string, memo: string) => {
 export const assertAndProcessMemo = (memo: string): string => {
   const [action] = memo.split(':')
 
+  if (!assertMemoHasAction(action, memo)) throw new Error(`action is required in memo: ${memo}`)
```
Comment (gomesalexandre) [inline src/lib/utils/thorchain/memo.ts:98]: And same here, the right side will never evaluate, we could instead return an empty string and add a similar comment to the others?

File: src/lib/utils/thorchain/memo.ts [hunk 3]
```diff
@@ -89,6 +115,8 @@ export const assertAndProcessMemo = (memo: string): string => {
     case 'a': {
       const [_action, pool, maybePairedAddr, , fee] = memo.split(':')
 
+      if (!pool) throw new Error(`pool is required in memo: ${memo}`)
```
Comment (gomesalexandre) [inline src/lib/utils/thorchain/memo.ts:118]:
```suggestion
      if (!assertMemoHasPool(pool, memo)) throw new Error(`pool is required in memo: ${memo}`)
```

Comment (gomesalexandre) [reply src/lib/utils/thorchain/memo.ts:118]: c437ee2da686cdb40220e4ec464f2dd3093eee20

File: src/lib/utils/thorchain/memo.ts (hunk 2 above)
Comment (gomesalexandre) [reply src/lib/utils/thorchain/memo.ts:98]: c437ee2da686cdb40220e4ec464f2dd3093eee20

File: src/lib/utils/thorchain/memo.ts [hunk 4]
```diff
@@ -2,43 +2,60 @@ import { bn } from 'lib/bignumber/bignumber'
 
 import { THORCHAIN_AFFILIATE_NAME } from './constants'
 
-const assertMemoHasPool = (pool: string, memo: string) => {
+const assertMemoHasPool = (pool: string | undefined, memo: string): pool is string => {
   if (!pool) throw new Error(`pool is required in memo: ${memo}`)
+  return true
 }
```
Comment (0xApotheosis) [inline src/lib/utils/thorchain/memo.ts:8]: We could use the type assertion syntax here, which I think is the more idiomatic way of doing this.
Then we could remove the `return true` as it'll do it implicitly so long as the type assertion function doesn't throw.

E.g.

```suggestion
const assertMemoHasPool = (pool: string | undefined, memo: string): asserts pool is string => {
  if (!pool) throw new Error(`pool is required in memo: ${memo}`)
}
```

File: src/lib/utils/thorchain/memo.ts (hunk 1 above)
Comment (0xApotheosis) [inline src/lib/utils/thorchain/memo.ts:52]: If we used the type assertion functions this line would just become something like:

```suggestion
assertMemoHasLimit(limit, memo)
```

**Review by 0xApotheosis — APPROVED** [review]
This is an improvement, and works functionally at runtime (https://viewblock.io/thorchain/tx/fc2b958de69cb5163fd940d64ef12dab315f90719cec6583651ed8247adc0d84).

Approving as is, but I think we should use the first-class assertion function syntax for these functions before merging.

File: src/lib/utils/thorchain/memo.ts (hunk 4 above)
Comment (gomesalexandre) [reply src/lib/utils/thorchain/memo.ts]: 8582fbd60c34e06b86104008523d3f7a69a26e54 - note the switch to good ol' function syntax: https://stackoverflow.com/a/72689922

File: src/lib/utils/thorchain/memo.ts (hunk 1 above)
Comment (gomesalexandre) [reply src/lib/utils/thorchain/memo.ts]: https://github.com/shapeshift/web/commit/8582fbd60c34e06b86104008523d3f7a69a26e54

**Pull request merged**